src/java/org/apache/cassandra/index/sai/iterators/KeyRangeIntersectionIterator.java

```diff
@@ -77,7 +77,19 @@ protected PrimaryKey computeNext()
                 if (index != alreadyAvanced)
                 {
                     KeyRangeIterator range = ranges.get(index);
-                    PrimaryKey nextKey = nextOrNull(range, highestKey);
+                    PrimaryKey nextKey = range.getCurrent();
+
+                    // Note that we will either have a data model that produces SKINNY primary keys or a data model
+                    // that produces some combination of WIDE and STATIC prikary keys.
+                    if (nextKey.kind() == PrimaryKey.Kind.WIDE || nextKey.kind() == highestKey.kind())
+                        // We can always skip if the target is of the same kind or this range is non-static. 
+                        nextKey = nextOrNull(range, highestKey);
+                    else if (nextKey.kind() == PrimaryKey.Kind.STATIC && nextKey.compareTo(highestKey) < 0)
+                        // For a range of static keys, only skip if we'e advanced to a new partition, and when we
+                        // do, skip to an actual static key. We may otherwise skip too far, as static row IDs always
+                        // precede non-static ones in on-disk postings lists.
+                        nextKey = nextOrNull(range, highestKey.toStatic());
```

Comment:
This is the core of the fix. Two problems:

1.) We were advancing streams of STATIC `PrimaryKey` objects from static column indexes before keys from the same partition from non-static column indexes could match. This was discovered by Harry on in-memory indexes.

2.) When we did skip the STATIC key streams, we could skip w/ a non-static `PrimaryKey` object, which would map to a posting row ID that was after the row ID of the static row, and so we could skip past the static `PrimaryKey` we wanted. (This affected on-disk indexes only.)

Reply:
The thing I'm most worried about here is static/static intersections and whether they terminate in all cases.

Reply:
It is not very obvious to me it always terminates. This `if / else if` has a theoretical path where it does *not* advance the range at all. Let's assume the first range returns a SKINNY key and the second range returns only STATIC keys then I guess the first range might get stuck because neither condition would be ever true. 

Im also not sure if now this invariant of the for loop in line 75 still holds:
```
// Once this inner loop finishes normally, all iterators are guaranteed to be at the same value.
```

Reply:
While I wasn't quite able to settle on a two-loop solution that I liked, I've tried to a.) further document our assumptions, b.) create some reasonable guardrails (see the new intersection compatibility check in `buildIterator()` and `Kind#isIntersectable()`), and c.) expand the randomized tests even further.

I'm convinced, at this point, that the patch is correct, and that even a rewrite of this intersection logic at a later time  would have sufficient tests ensuring it wouldn't make the same mistakes :)

Let me know what you think...

Reply:
I am quite convinced your patch is correct, but I managed to find a way to make the conditions in the main loop simpler (and still leaving only one loop). Now there is exactly one condition to decide whether we bump up the iterator, and another one to decide how we advance (so we advance to the right partition in case of static column):

https://github.com/apache/cassandra/commit/b8b15e2a7c9c0a635ab40bcd9992476f0aa37c25